services/motion/builtin/builtin.go

```diff
@@ -237,71 +247,170 @@ func (ms *builtIn) MoveOnGlobe(
+	obstaclePollingPeriod := time.Duration(1000/motionCfg.ObstaclePollingFreqHz) * time.Millisecond
 
-	if motionCfg.LinearMPerSec != 0 {
-		kinematicsOptions.LinearVelocityMMPerSec = motionCfg.LinearMPerSec * 1000
-	}
-	if motionCfg.AngularDegsPerSec != 0 {
-		kinematicsOptions.AngularVelocityDegsPerSec = motionCfg.AngularDegsPerSec
-	}
-	if motionCfg.PlanDeviationM != 0 {
-		kinematicsOptions.PlanDeviationThresholdMM = motionCfg.PlanDeviationM * 1000
+	movementSensor, ok := ms.movementSensors[movementSensorName]
+	if !ok {
+		return false, resource.DependencyNotFoundError(movementSensorName)
 	}
-	kinematicsOptions.GoalRadiusMM = math.Min(motionCfg.PlanDeviationM*1000, 3000)
-	kinematicsOptions.HeadingThresholdDegrees = 8
 
-	plan, kb, err := ms.planMoveOnGlobe(
-		ctx,
-		componentName,
-		destination,
-		movementSensorName,
-		obstacles,
-		kinematicsOptions,
-		extra,
-	)
+	planRequest, kb, err := ms.newMoveOnGlobeRequest(ctx, componentName, destination, movementSensor, obstacles, motionCfg, extra)
 	if err != nil {
-		return false, fmt.Errorf("error making plan for MoveOnMap: %w", err)
+		return false, err
 	}
-	// execute the plan
-	for i := 1; i < len(plan); i++ {
-		ms.logger.Info(plan[i])
-		if inputEnabledKb, ok := kb.(inputEnabledActuator); ok {
-			if err := inputEnabledKb.GoToInputs(ctx, plan[i]); err != nil {
-				// If there is an error on GoToInputs, stop the component if possible before returning the error
-				if stopErr := kb.Stop(ctx, nil); stopErr != nil {
-					return false, errors.Wrap(err, stopErr.Error())
+
+	successChan := make(chan bool)
+	defer close(successChan)
+
+	replanChan := make(chan bool, 1)
+	defer close(replanChan)
+	replanChan <- true
+
+	errChan := make(chan error)
+	defer close(errChan)
+
+	var backgroundWorkers sync.WaitGroup
+	defer backgroundWorkers.Wait()
+
+	// helper function to manage polling functions
+	startPolling := func(ctx context.Context, period time.Duration, fn func(context.Context) error) {
+		backgroundWorkers.Add(1)
+		goutils.ManagedGo(func() {
+			ticker := time.NewTicker(period)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					if err := fn(ctx); err != nil {
+						errChan <- err
+					}
+					return
 				}
-				return false, err
 			}
+		}, backgroundWorkers.Done)
+	}
+
+	var cancelCtx context.Context
+	var cancelFn context.CancelFunc
+	defer func() {
+		if cancelFn != nil {
+			cancelFn()
+		}
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		case <-successChan:
+			return true, nil
+		case err := <-errChan:
+			return false, err
+		case <-replanChan:
+			fmt.Println("replanning")
+
+			if cancelFn != nil {
+				cancelFn()
+			}
+			backgroundWorkers.Wait()
+			cancelCtx, cancelFn = context.WithCancel(ctx)
+
+			inputs, err := kb.CurrentInputs(ctx)
+			if err != nil {
+				return false, nil
+			}
+			if len(kb.Kinematics().DoF()) == 2 {
```

Comment:
This is awkward and hacky and only coincidentally working right now.
It should have 1) an explicit type check, 2) lots of comments, and 3) probably a logger message mentioning this is happening. Logger could be upstream so it only triggers once.

Reply:
We need to think harder about how to handle this correctly and I would rather not do it here.  I'm going to add a TODO and I hope youre ok with pushing this improvement out until later

Reply:
I'm fine with a TODO on account of it not actually being possible for any external user's frame to be passed in here.